### ISSUE TITLE
is there an install syntax to specify package version?

### ISSUE THREAD

**Issue by fangq** [msys2/setup-msys2#287]
I would trying to install a specific version of gcc (gcc-9) in an [action setting](https://github.com/fangq/mmc/blob/2b11b33788042f815925e41318f700ff181dbad2/.github/workflows/build_win64.yml#L29-L37), but could not find any examples to do so - is this supported?

additionally, I am looking for the package that provides static library `libgomp.a`, I only found that [`mingw-w64-x86_64-gcc-libs`](https://packages.msys2.org/package/mingw-w64-x86_64-gcc-libs) provides `libgomp.so`, but not the static library.

any help is appreciated!

**Comment by Biswa96** [comment]
It is _not_ possible to install old packages like that. Old packages can be found in https://mirror.msys2.org/mingw/ repository. But installing them may break the whole setup and you have to manually install those packages with `pacman -U <full-link-here>` command.

**Comment by Biswa96** [comment]
> I would trying to install a specific version of gcc (gcc-9) in an [action setting](https://github.com/fangq/mmc/blob/master/.github/workflows/build_win64.yml#L38-L46)

In the actions log, there seems to be some path conflict between the mingw toolchains in chocolatey and msys2. If you have any issue with latest gcc toolchain it may be fixed faster than using old gcc.

**Comment by fangq** [comment]
thanks for your quick reply and helpful comments. I realized that `pacman` and `msys2` was made for rolling releases, which is aimed to only serve the latest.

> In the actions log, there seems to be some path conflict between the mingw toolchains in chocolatey and msys2. If you have any issue with latest gcc toolchain it may be fixed faster than using old gcc.

I agree. The actual issue I am trying to solve is the missing of static libraries: `libz.a`, `libgomp.a` and `libpthread.a` when I used the default Mingw64 8.1 gcc toolchain provided by the windows-2019 runner. here is the log

https://github.com/fangq/mmc/actions/runs/4356395089/jobs/7614344016

I searched and only found commands to install packages via msys2, if I am able to locate the packages in msys2 that provides these .a files, I should be able to set `shell: msys2` and use the mingw64 gcc installed on msys2 (12.x) to complete the compilation;

I just tried to mix the mingw64 gcc (9.0.2) with the .a files extracted from msys2, but it failed with undefined reference errors

https://github.com/fangq/mmc/actions/runs/4357600651/jobs/7617111580

I thought that if I could downgrade the msys2 gcc version, I may have a better luck to link.

alternatively, I may be able to solve this by finding an action script to install libraries for mingw64 instead of msys2, but I failed to find such command :-(

**Comment by Biswa96** [comment]
> if I am able to locate the packages in msys2 that provides these .a files

You can search files with pacman. For example, here I am searching for libpthread.a in ucrt64

```
$ pacman -Fq libpthread.a | grep ucrt
ucrt64/mingw-w64-ucrt-x86_64-winpthreads-git
```

* libpthread.a: https://packages.msys2.org/base/mingw-w64-winpthreads-git
* libz.a: https://packages.msys2.org/base/mingw-w64-zlib
* libgomp.a: https://packages.msys2.org/base/mingw-w64-gcc

**Comment by fangq** [comment]
> In the actions log, there seems to be some path conflict between the mingw toolchains in chocolatey and msys2. If you have any issue with latest gcc toolchain it may be fixed faster than using old gcc.

@Biswa96, you were spot-on! turned out most of my struggle was a result of mixing gcc versions - I added path to Octave's bin folder which also contains a bundled gcc. After appending mingw64's gcc path in GITHUB_ENV, I was able to run static linking without needing to install any additional package (or msys2 environment).

closing the ticket. thanks again for all the helpful comments!
